driver/main.c

```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
         if (dist < min_dist) {
-          min_dist                          = dist;
-          hetero_rain->data2mesh_idx[icell] = idx;
+          min_dist                   = dist;
+          data->data2mesh_idx[icell] = idx;
         }
         idx++;
       }
     }
 
     if (0) {
-      PetscInt idx = hetero_rain->data2mesh_idx[icell];
-      printf("%04" PetscInt_FMT " %f %f %02" PetscInt_FMT " %f %f\n", icell, xc, yc, idx, hetero_rain->data_xc[idx], hetero_rain->data_yc[idx]);
+      PetscInt idx = data->data2mesh_idx[icell];
+      printf("%04" PetscInt_FMT " %f %f %02" PetscInt_FMT " %f %f\n", icell, xc, yc, idx, data->data_xc[idx], data->data_yc[idx]);
     }
   }
 
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-// set spatially heterogeneous rainfall rate
-PetscErrorCode SetHeterogeneousRainfall(HeterogeneousRainData *hetero_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
+/// @brief Set rainfall rate for local cells from a spatially raster rainfall dataset. It is assumed that
+///        rainfall rate is mm/hr.
+/// @param data     [in]  A pointer to RasterDataset struct
+/// @param cur_time [in]  Current time
+/// @param ncells   [in]  Number of local cells
+/// @param *rain    [out] Rainfall rate for local cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetRasterDataset(RasterDataset *data, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
   PetscFunctionBegin;
 
   // Is it time to open a new file?
-  if (cur_time / 3600.0 >= (hetero_rain->ndata_file) * hetero_rain->dtime_in_hour) {
-    OpenANewHeterogeneousRainfallData(hetero_rain);
+  if (cur_time / 3600.0 >= (data->ndata_file) * data->dtime_in_hour) {
+    OpenANewRasterDataset(data);
   }
 
-  PetscInt  offset                = hetero_rain->header_offset;
+  PetscInt  offset                = data->header_offset;
   PetscReal mm_per_hr_2_m_per_sec = 1.0 / (1000.0 * 3600.0);
 
   for (PetscInt icell = 0; icell < ncells; icell++) {
-    PetscInt idx = hetero_rain->data2mesh_idx[icell];
-    rain[icell]  = hetero_rain->data_ptr[idx + offset] * mm_per_hr_2_m_per_sec;
+    PetscInt idx = data->data2mesh_idx[icell];
+    rain[icell]  = data->data_ptr[idx + offset] * mm_per_hr_2_m_per_sec;
   }
 
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
-// set spatially homogeneous rainfall rate for all grid cells
-PetscErrorCode SetHomogeneousRainfall(HomogeneousRainData *homogeneous_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
+/// @brief Sets rainfall value for local grid cells based on spatially homogenous, temporally-varying rainfall dataset
+/// @param homogeneous_rain [in]  A pointer to HomogeneousDataset
+/// @param cur_time         [in]  Current time
+/// @param ncells           [in]  Number of local cells
+/// @param *rain            [out] Rainfall rate for local cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetHomogeneousRainfall(HomogeneousDataset *homogeneous_rain, PetscReal cur_time, PetscInt ncells, PetscReal *rain) {
```

Comment:
We could use `SetHomogeneousData` here, analogous to `Set{Unstructured,Raster}Data` above.